README.md

```diff
@@ -41,76 +34,77 @@ The current set of features include:
+        sudo apt install default-jdk
 
-Make sure you have Gradle 3.2 (or later) installed. Ubuntu 16.04 ships with Gradle 2.10, you can install a more recent version of Gradle from the following PPA:
+1. Install Gradle.
+Make sure you have Gradle 3.2 (or later) installed.
 
-```
-$ sudo add-apt-repository ppa:cwchien/gradle
-$ sudo apt install -y gradle
-```
-For OSX, use homebrew command to install gradle:
+    *Ubuntu Bionic or later*
 
-```
-$ brew install gradle
-```
+        sudo apt install gradle
 
-For OSX users, there some compatibily issues have been reported between gradle 3.x and Java 9. Currently only the combination of Gradle 3.x + Java 8 has been tested successfully. Make sure to use install and use Java 8 for the building process:
+    *macOS*
 
-```
-$ brew tap caskroom/versions
-$ brew cask install java8
-$ export JAVA_HOME=/Library/Java/JavaVirtualMachines/1.8.0.jdk/Contents/Home
-``` 
+        brew install gradle
 
-> Note: On Windows, you may use `choco install -y gradle`
+    Note: if run into compatibily issues between gradle 3.x and Java 9, try using Java 8,
 
-The following sections deal with building the `ros2_java` codebase for the desktop Java runtime and for Android.
+        brew tap caskroom/versions
+        brew cask install java8
+        export JAVA_HOME=/Library/Java/JavaVirtualMachines/1.8.0.jdk/Contents/Home
 
-Desktop
--------
+    *Windows*
 
-```
-mkdir -p ros2_java_ws/src
-cd ros2_java_ws
-curl -skL https://raw.githubusercontent.com/ros2-java/ros2_java/dashing/ros2_java_desktop.repos -o ros2_java_desktop.repos
-vcs import src < ros2_java_desktop.repos
-. ../ament_ws/install_isolated/local_setup.sh
-ament build --symlink-install --isolated
-```
+        choco install gradle
+
+1. Install build tools:
+
+        sudo apt install curl python3-colcon-common-extensions python3-pip python3-vctool
+
+1. Install Gradle extensions for colcon:
+
+        python3 -m pip install -U git+https://github.com/colcon/colcon-gradle
+        python3 -m pip install --no-deps -U git+https://github.com/colcon/colcon-ros-gradle
+
+### Download and Build ROS 2 Java for Desktop
+
+1. Source your ROS 2 installation, for example:
+
+        source /opt/ros/dashing/setup.bash
+
+1. Download the ROS 2 Java repositories into a workspace:
+
+        mkdir -p ros2_java_ws/src
+        cd ros2_java_ws
+        curl -skL https://raw.githubusercontent.com/ros2-java/ros2_java/dashing/ros2_java_desktop.repos | vcs import src
+
+1. **Linux only** Install ROS dependencies:
+
+        rosdep install --from-paths src -y -i --skip-keys "ament_tools"
+
+1. Build desktop packages:
 
-> On Windows, if you would like to use OpenSplice, call `call "C:\opensplice67\HDE\x86_64.win64\release.bat"` before building.
+        colcon build --symlink-install
 
-Now you can just run a bunch of examples, head over to https://github.com/esteve/ros2_java_examples for more information.
+### Download and Build ROS 2 Java for Android
```

Comment:
I prefer not to merge this PR until the Android instructions are complete, some of the users that use ros2-java use it only because of the Android support.

Reply:
I'd argue that this isn't making things worse for Android, it just isn't improving it either.  So I'd still say we should merge this one (just so it doesn't linger), and then when somebody gets around to updating Android it would go on top of this.

Reply:
> I'd argue that this isn't making things worse for Android, it just isn't improving it either.

It definitely make things worse for Android users, because they won't have a path for building `ros2_java`. Currently they can build `ros2_java` for Bouncy, but once `master` only works for Dashing, they won't have instructions for building `ros2_java`.

> and then when somebody gets around to updating Android it would go on top of this.

The goal of `ros2_java` is to support JRE and Android, so any contribution must target both.

Reply:
> It definitely makes things worse for Android users, because they won't have a path for building `ros2_java`. Currently they can build `ros2_java` for Bouncy, but once `master` only works for Dashing, they won't have instructions for building `ros2_java`.

I guess my main point is that this PR targets the `dashing` branch, which already doesn't work for Android.  So we can merge it onto the dashing branch; we just can't merge the dashing branch onto master.

Reply:
I guess this raises a separate question about organization. I'd been working under the idea that we would maintain separate branches for each ROS distribution (like many other ROS packages). If users want to build on top of Bouncy, then they can refer to the [`bouncy`](https://github.com/ros2-java/ros2_java/tree/bouncy) branch.

To make it more clear for users, we could add a note near the top of the README referring them to the correct branch (with instructions) for their target distribution.

Reply:
In the end, my qualm is that OSRF has its own agenda regarding `ros2-java`, which seems to be aligned for the most part with the goals I have when I started this project, except for the Android stuff. But this is not an OSRF project, so all contributors are treated equally, no matter who their employer is.

To be honest, I'm less than enthusiastic about OSRF seemingly trying to set the goals for `ros2-java`, hence why I prefer to keep this project separate from them. I feel strongly about supporting Android, and any contribution must take Android into account, but you guys seem very keen on not updating `ros2-java` for Android in your contributions, whether code or documentation. Frankly, updating the README with instructions for Android requires much less work than setting up different branches for each distro.

Reply:
It's true that for our use-cases we don't have as strong an interest in Android, but I do think we should still support it. Before opening this PR I did try to compile rcljava for Android to update the instructions and add CI, but haven't had any success yet. I'm not sure what needs to change in the documentation (or code) to make the instructions work. Unfortunately, since it's not a priority it's difficult for me to spend more time on it at the moment (hence the TODO). We can wait to merge this until we figure it out, but I thought it would be nicer to get the current changes in as an incremental improvement.

Regarding branches for each distro, I guess it's a matter of opinion. I think it makes maintenance and development a lot easier. Unless we only want to support the latest distro, we're bound to make changes to the default branch that breaks compatibility with older distros. If we want to maintain compatibility on a single branch, then we have to do things like "ifdef" around code blocks, which seems less pleasant (IMO) than maintaining distro-specific branches. I've tried to structure code changes such that `dashing` will remain compatible with Dashing and then I'd like to cherry-pick changes I have on my [dev branch](https://github.com/ros2-java/ros2_java/tree/jacob/foxy) for Foxy compatibility to `master` (since the changes to Foxy are not compatibility with Dashing). I'm open to other strategies for maintaining compatibility between distros. I shouldn't have presumed that we'd use this branching method, sorry about that.

Reply:
> In the end, my qualm is that OSRF has its own agenda regarding `ros2-java`, which seems to be aligned for the most part with the goals I have when I started this project, except for the Android stuff. But this is not an OSRF project, so all contributors are treated equally, no matter who their employer is.

Sure.  That only makes sense; every contributor has their own agenda for contributing to the project.

> 
> To be honest, I'm less than enthusiastic about OSRF seemingly trying to set the goals for `ros2-java`, hence why I prefer to keep this project separate from them. I feel strongly about supporting Android, and any contribution must take Android into account, but you guys seem very keen on not updating `ros2-java` for Android in your contributions, whether code or documentation. Frankly, updating the README with instructions for Android requires much less work than setting up different branches for each distro.

We are just setting the goals for the things that we are interested in.  As you can probably tell, we aren't much interested in Android.  I think the disconnect we currently have is that there are 2 main branches of development:

* Bouncy with Android support
* Dashing and later without Android support

Jacob and I are mostly interested in the latter; you are mostly interested in the former.

It's not that we don't care at all about the Android goal of the project, but it isn't clear to me how to keep that support working.  Since it isn't our main target, its hard to justify spending a lot of time setting it up.

There are several ways to rectify the disconnect:
1.  Provide a docker image to automatically build/test Android support
1.  Keep developing on separate branches, one for Android, one for other things
1.  Fork the repository

There may be others.

Reply:
> We are just setting the goals for the things that we are interested in.

OSRF can not set the goals for the project, basically because they are not the owners of `ros2_java`

> It's not that we don't care at all about the Android goal of the project, but it isn't clear to me how to keep that support working. Since it isn't our main target, its hard to justify spending a lot of time setting it up.

I don't think it's that much time. I'm more than happy to help you if you have problems with setting it up

>  I think the disconnect we currently have is that there are 2 main branches of development:
> 
>     Bouncy with Android support
>     Dashing and later without Android support

Maintaining multiple branches requires a lot of effort, which in this case is completely unnecessary because there's nothing in the code that prevents both platforms from being supported. My goal is to only have one branch, it's much simpler that way

>    Keep developing on separate branches, one for Android, one for other things

The Android code is isolated in other layers in ROS 2, there's nothing in `ros2_java`, either in master or the dashing branch that prevents it from running on Android, only the instructions need to be updated.

>    Fork the repository

Whoa, the nuclear option just because you don't want to update a README. Frankly, not a big fan of this level of confrontation

> There may be others.

When contributing to a project you are not even a committer, it's good etiquette to try to cooperate at least, at least from my experience

Reply:
I've pushed changes to the instructions that I've tried so far and opened a PR targeting this branch https://github.com/ros2-java/ros2_java/pull/125

I can continue to chip away at it in some free time, but any insights would be greatly appreciated :slightly_smiling_face:

Reply:
I've merged #125 here, containing instructions for Android (and CI!) (a3fa46d)